### PR TITLE
Adds Managed to Native abort logic.

### PR DESCRIPTION
**Pull request by jkotalik**

### CODE REVIEW

**Comment by jkotalik** [comment]
Abort is a bit trickier here because after notifying IIS to abort, reads and writes will not fail by calling ReadEntityBody/WriteEntityChunks. Therefore, I add aborted to the stored context. @pan-wang and I will go through the native to managed cancellation later.

File: src/AspNetCore/Src/inprocessstoredcontext.cxx [hunk 1]
```diff
@@ -32,6 +32,22 @@ IN_PROCESS_STORED_CONTEXT::QueryHttpContext(
     return m_pHttpContext;
 }
 
+VOID
+IN_PROCESS_STORED_CONTEXT::SetRequestAborted(
+    BOOL fRequestAborted
```
Comment (pan-wang) [inline src/AspNetCore/Src/inprocessstoredcontext.cxx]: since you cannot change a request  back to un-aborted if it got aborted, no need for passing in the bool value

File: src/AspNetCore/Src/inprocessapplication.cxx [hunk 2]
```diff
@@ -176,12 +201,26 @@ http_write_response_bytes(
     _In_ BOOL* pfCompletionExpected
 )
 {
+    HRESULT hr;
```
Comment (pan-wang) [inline src/AspNetCore/Src/inprocessapplication.cxx]: please initialize it to S_OK

File: src/AspNetCore/Src/inprocessapplication.cxx [hunk 3]
```diff
@@ -176,12 +201,26 @@ http_write_response_bytes(
...
+        return hr;
+    }
+
+    if (pInProcessStoredContext->QueryRequstAborted())
+    {
+        *pfCompletionExpected = FALSE;
+        return E_ABORT;
+    }
+
     IHttpResponse *pHttpResponse = (IHttpResponse*)pHttpContext->GetResponse();
```
Comment (pan-wang) [inline src/AspNetCore/Src/inprocessapplication.cxx]: please declare the variables at the top of the function

File: src/AspNetCore/Src/inprocessapplication.cxx [hunk 4]
```diff
@@ -46,13 +46,26 @@ http_get_raw_response(
     return pHttpContext->GetResponse()->GetRawHttpResponse();
 }
 
-EXTERN_C __MIDL_DECLSPEC_DLLEXPORT VOID http_set_response_status_code(
+EXTERN_C __MIDL_DECLSPEC_DLLEXPORT HRESULT http_set_response_status_code(
```
Comment (pan-wang) [inline src/AspNetCore/Src/inprocessapplication.cxx]: the signature may need to updated as you may also want to set sub-status 
https://support.microsoft.com/en-us/help/943891/the-http-status-code-in-iis-7-0--iis-7-5--and-iis-8-0

Comment (jkotalik) [reply src/AspNetCore/Src/inprocessapplication.cxx]: This is fixed, diff still showing up.

File: src/AspNetCore/Src/inprocessapplication.cxx [hunk 5]
```diff
@@ -46,13 +46,27 @@ http_get_raw_response(
...
+    {
+        return hr;
+    }
+
+    if (pInProcessStoredContext->QueryRequstAborted())
+    {
+        return E_ABORT;
+    }
+
+    return pHttpContext->GetResponse()->SetStatus(statusCode, pszReason, subStatusCode);
```
Comment (Tratcher) [inline src/AspNetCore/Src/inprocessapplication.cxx]: What would happen if you called SetStatus after ResetConnection? How about WriteEntityChunks? Does it matter?
I wonder if we can skip all of this new API and call ResetConnection directly from managed.

File: src/AspNetCore/Src/inprocessstoredcontext.cxx [hunk 6]
```diff
@@ -32,6 +32,21 @@ IN_PROCESS_STORED_CONTEXT::QueryHttpContext(
...
 
+VOID
+IN_PROCESS_STORED_CONTEXT::SetRequestAborted(
+)
+{
+    m_fRequestAborted = TRUE;
+}
+
+BOOL
+IN_PROCESS_STORED_CONTEXT::QueryRequstAborted(
```
Comment (Tratcher) [inline src/AspNetCore/Src/inprocessstoredcontext.cxx]: Request

File: src/AspNetCore/Src/inprocessapplication.cxx (hunk 5 above)
Comment (jkotalik) [reply src/AspNetCore/Src/inprocessapplication.cxx]: Calling SetStatus and WriteEntityChunks will succeed after calling ResetConnection. It is just being written to a local buffer. 
@pan-wang as part of the redesign of the pInvokes, will we be directly exposing HttpContext methods? If so, I'd agree it is better to have many of these calls in managed as currently most are wrappers for the native calls.

File: src/AspNetCore/Src/inprocessapplication.cxx [hunk 7]
```diff
@@ -46,13 +46,27 @@ http_get_raw_response(
...
-    pHttpContext->GetResponse()->SetStatus(statusCode, pszReason);
+    HRESULT hr S_OK;
+    IN_PROCESS_STORED_CONTEXT* pInProcessStoredContext = NULL;
+    hr = IN_PROCESS_STORED_CONTEXT::GetInProcessStoredContext(pHttpContext, &pInProcessStoredContext);
+    if (FAILED(hr))
+    {
+        return hr;
+    }
+
+    if (pInProcessStoredContext->QueryRequestAborted())
```
Comment (pan-wang) [inline src/AspNetCore/Src/inprocessapplication.cxx]: current design looks a lit bit wired. The request is aborted by the managed layer. Why not immediately stop reading the request at the managed layer once the managed layer decides to abort the request instead of continue reading and checking it at the native layer.

Comment (jkotalik) [reply src/AspNetCore/Src/inprocessapplication.cxx]: Yeah I actually think I can remove these checks entirely. Writes and reads, if they occur, should succeed. The request will truly be aborted when the disconnect listener is fired.

File: src/AspNetCore/Src/managedexports.cxx [hunk 8]
```diff
@@ -325,4 +339,15 @@ http_cancel_io(
     return pHttpContext->CancelIo();
 }
 
+EXTERN_C __MIDL_DECLSPEC_DLLEXPORT
+HRESULT
+http_abort_request(
```
Comment (davidfowl) [inline src/AspNetCore/Src/managedexports.cxx]: http_reset_connection

File: src/AspNetCore/Src/managedexports.cxx [hunk 9]
```diff
@@ -173,12 +177,22 @@ http_write_response_bytes(
     _In_ BOOL* pfCompletionExpected
 )
 {
-    IHttpResponse *pHttpResponse = (IHttpResponse*)pHttpContext->GetResponse();
+    HRESULT hr S_OK;
+    IHttpResponse *pHttpResponse = NULL;
+    IN_PROCESS_STORED_CONTEXT* pInProcessStoredContext = NULL;
```
Comment (pan-wang) [inline src/AspNetCore/Src/managedexports.cxx]: no one reference to this stored context

**Comment by jkotalik** [comment]
This will be merged pending #212

File: src/AspNetCore/Src/managedexports.cxx [hunk 10]
```diff
@@ -158,19 +159,22 @@ http_read_request_bytes(
     _Out_ BOOL* pfCompletionPending
 )
 {
-    HRESULT hr;
+    HRESULT hr S_OK;
+    IN_PROCESS_STORED_CONTEXT* pInProcessStoredContext = NULL;
```
Comment (davidfowl) [inline src/AspNetCore/Src/managedexports.cxx]: What is this for?

File: src/AspNetCore/Src/managedexports.cxx [hunk 11]
```diff
@@ -158,19 +159,22 @@ http_read_request_bytes(
...
     if (pHttpContext == NULL)
     {
         return E_FAIL;
     }
+
     if (dwCbBuffer == 0)
     {
         return E_FAIL;
     }
-    IHttpRequest *pHttpRequest = (IHttpRequest*)pHttpContext->GetRequest();
```
Comment (davidfowl) [inline src/AspNetCore/Src/managedexports.cxx:171]: What's up with these changes? They look unrelated?

Comment (jkotalik) [reply src/AspNetCore/Src/managedexports.cxx:171]: This is just style in keeping the declarations at the beginning of the file.

**Comment by jkotalik** [comment]
@pan-wang Let's coordinate this PR with the disconnect listener changes.

File: src/AspNetCore/Src/managedexports.cxx [hunk 12]
```diff
@@ -197,12 +200,21 @@ http_write_response_bytes(
     _In_ BOOL* pfCompletionExpected
 )
 {
-    IHttpResponse *pHttpResponse = (IHttpResponse*)pHttpContext->GetResponse();
+    HRESULT hr S_OK;
```
Comment (pan-wang) [inline src/AspNetCore/Src/managedexports.cxx:203]: where is "="?

File: src/AspNetCore/Src/managedexports.cxx [hunk 13]
```diff
@@ -197,12 +200,21 @@ http_write_response_bytes(
...
+    IHttpResponse *pHttpResponse = NULL;
     BOOL fAsync = TRUE;
     BOOL fMoreData = TRUE;
     DWORD dwBytesSent = 0;
 
-    HRESULT hr = pHttpResponse->WriteEntityChunks(
+
+    if (pHttpContext == NULL)
+    {
+        return E_FAIL;
```
Comment (pan-wang) [inline src/AspNetCore/Src/managedexports.cxx:212]: This null check is questionable as we control the stack unless you intentionally allow null input. Otherwise you can put null debug_assert to indicate that null input should not allowed.

this function can be simplified

File: src/AspNetCore/Src/managedexports.cxx [hunk 14]
```diff
@@ -221,12 +233,12 @@ http_flush_response_bytes(
 )
 {
     IHttpResponse *pHttpResponse = (IHttpResponse*)pHttpContext->GetResponse();
-
+    HRESULT hr = S_OK;
```
Comment (pan-wang) [inline src/AspNetCore/Src/managedexports.cxx:236]: since you already hardcoded the value, why do you still need those local variables

File: src/AspNetCore/Src/managedexports.cxx [hunk 15]
```diff
@@ -349,4 +361,15 @@ http_cancel_io(
     return pHttpContext->CancelIo();
 }
 
+EXTERN_C __MIDL_DECLSPEC_DLLEXPORT
+HRESULT
+http_reset_connection(
+    _In_ IHttpContext* pHttpContext
```
Comment (pan-wang) [inline src/AspNetCore/Src/managedexports.cxx:367]: not sure about why you want to return hardcoded hresult

**Comment by jkotalik** [comment]
Closing for #227
